syntax = "proto2";

package detection.protos;

message Optimizer {
  oneof optimizer_oneof {
    RMSPropOptimizer rms_prop_optimizer = 1;
    MomentumOptimizer momentum_optimizer = 2;
    AdamOptimizer adam_optimizer = 3;
  }

  optional bool use_moving_average = 4 [default = true];
  optional float moving_average_decay = 5 [default = 0.9999]; 
}

message RMSPropOptimizer {
  optional LearningRate learning_rate = 1;
  optional float momentum_optimizer_value = 2 [default = 0.9];
  optional float decay = 3 [default = 0.9];
  optional float epsilon = 4 [default = 1.0];
}

message MomentumOptimizer {
  optional LearningRate learning_rate = 1;
  optional float momentum_optimizer_value = 2 [default = 0.9];
}

message AdamOptimizer {
  optional LearningRate learning_rate = 1;
}

message LearningRate {
  oneof learning_rate_oneof {
    ConstantLearningRate constant_learning_rate = 1;
    ExponentialDecayLearningRate exponential_decay_learning_rate = 2;
  }
}

message ConstantLearningRate {
  optional float learning_rate = 1 [default = 0.002];
}

message ExponentialDecayLearningRate {
  optional float initial_learning_rate = 1 [default = 0.002];
  optional uint32 decay_steps = 2 [default = 4000000];
  optional float decay_factor = 3 [default = 0.95];
  optional bool staircase = 4 [default = true];
}

